playground/engine.py

```diff
@@ -0,0 +1,818 @@
+        rays = self.raygen(camera, use_spp=is_use_spp)
+
+        if is_first_pass:
+            if not self.primitives.enabled or not self.primitives.has_visible_objects():
+                rb = self.scene_mog.trace(rays_o=rays.rays_ori, rays_d=rays.rays_dir)
+            else:
+                rb = self._render_playground_hybrid(rays.rays_ori, rays.rays_dir)
+
+            rb = dict(rgb=rb['pred_rgb'], opacity=rb['pred_opacity'])
+            rb['rgb'] = torch.pow(rb['rgb'], 1.0 / self.gamma_correction)
+            rb['rgb'] = rb['rgb'].mean(dim=0).unsqueeze(0)
+            rb['opacity'] = rb['opacity'].mean(dim=0).unsqueeze(0)
+            self.spp.reset_accumulation()
+            self.depth_of_field.reset_accumulation()
+        else:
+            # Render accumulated effects, i.e. depth of field
+            rb = dict(rgb=self.last_state['rgb_buffer'], opacity=self.last_state['opacity'])
+            if self.use_depth_of_field:
+                self._render_depth_of_field_buffer(rb, camera, rays)
+            elif self.use_spp:
+                self._render_spp_buffer(rb, rays)
+
+        # Keep a noisy version of the accumulated rgb buffer so we don't repeat denoising per frame
+        rb['rgb_buffer'] = rb['rgb']
+        if self.use_optix_denoiser:
+            rb['rgb'] = self.tracer.denoise(rb['rgb'])
+
+        if rays.mask is not None:  # mask is for masking away pixels out of view for, i.e. fisheye
+            mask = rays.mask[None, :, :, 0]
+            rb['rgb'][mask] = 0.0
+            rb['rgb_buffer'][mask] = 0.0
+            rb['opacity'][mask] = 0.0
+
+        self.cache_last_state(camera=camera, renderbuffers=rb, canvas_size=[camera.height, camera.width])
+        return rb
+
+    @torch.cuda.nvtx.range("render")
+    def render(self, camera: Camera) -> Dict[str, torch.Tensor]:
+        """
+        Renders a single frame, possibly consisting of multiple passes if any visual effects are toggled on.
+        render() is suitable for offline rendering situations, where high quality is preferred.
+
+        By default, 3dgrt requires only a single pass to render.
+        Toggling on antialiasing and depth of field may require additional samples which require additional passes.
+
+        The returned dictionary of rendered buffers always includes the 'rgb' and 'opacity' buffers.
+        """
+        renderbuffers = self.render_pass(camera, is_first_pass=True)
+        while self.has_progressive_effects_to_render():
+            renderbuffers = self.render_pass(camera, is_first_pass=False)
+        return renderbuffers
+
+    def invalidate_materials_on_gpu(self):
+        """ Marks the materials on GPU as out of date.
+        Materials and textures will be uploaded to the GPU during the next rendering pass.
+        """
+        self.is_materials_dirty = True
+
+    @torch.cuda.nvtx.range("load_3dgrt_object")
+    def load_3dgrt_object(self, object_path, config_name='apps/colmap_3dgrt.yaml'):
```

Comment:
please use smarter path handling for configs. Some ideas:
* if training copies its config, could look for config there
* if config_name is relative, could search in path relative to this file (so users can call class from other dirs)